feat: effect provenance

gaganchandan opened the pull request:
**WIP**: First iteration of effect provenance discussed in #10785 using the new `Source` provenance.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/typer/EffectProvenance.scala:
```diff
@@ -0,0 +1,198 @@
+  // provenance of type Provenance.ExpectEffect and an effect in the
+  // in the concrete set is not in the expected effects.
+  def debug(constrs0: List[TypeConstraint]): Unit = {
+    if (enableEffectProvenance) {
```
Can we break some of this code down into smaller helper functions? Preferably in a functional style.

mlutze replied:
Let's invert the if and turn this into a guard.

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/EffectProvenance.scala:
```diff
@@ -0,0 +1,198 @@
+  private def printConstrs(constrs: List[TypeConstraint]): Unit = {
+    constrs.foreach {
+      case TypeConstraint.Equality(eff1, eff2, prov) => prov match {
+        case _ => println(s"$eff1 ~ $eff2 at ${prov.loc} with provenance $prov")
```
Unneeded match?

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/EffectProvenance.scala:
```diff
@@ -0,0 +1,198 @@
+                      case Type.Pure => findPath(eff1, Symbol.mkEffectSym("Pure"), constr :: Nil, constrs0.filterNot(_ == constr))
+                      case _ => {
+                        val (add, _) = findConcreteEffects(eff2)
+                        eff2.effects.filter(add.contains(_)).foreach {
```
A couple inline comments here would be nice, since this is a bit complex.

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/EffectProvenance.scala:
```diff
@@ -0,0 +1,198 @@
+                (eff1, eff2) match {
+                  case _ => {
+                    eff2 match {
+                      case Type.Pure => findPath(eff1, Symbol.mkEffectSym("Pure"), constr :: Nil, constrs0.filterNot(_ == constr))
```
Making the effect symbol `Pure` seems brittle.

Would it make sense instead to create an `Option`-like type?

```scala
sealed trait XYZ

object XYZ {
    case class Effect(effectSym: EffSym) extends XYZ
    case class Pure extends XYZ
}
```

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/EffectProvenance.scala:
```diff
@@ -0,0 +1,198 @@
+                  }
+                }
+              }
+              case _ => ()
```
These dangling wildcard cases are hard to match up. Can we simplify by introducing one or more helper functions?

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/EffectProvenance.scala:
```diff
@@ -0,0 +1,198 @@
+                    tpeVar match {
+                      case Type.Var(_, _) =>
+                        findPath(tpeVar, concreteEff, constr :: path, constrs.filterNot(_ == constr))
+                      case _ => ()
```
This case cannot happen, right?

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/EffectProvenance.scala:
```diff
@@ -0,0 +1,198 @@
+  // in the concrete set is not in the expected effects.
+  def debug(constrs0: List[TypeConstraint]): Unit = {
+    if (enableEffectProvenance) {
+      constrs0.foreach { constr =>
+        constr match {
```
Scala allows simplifying functions where we immediately match on the argument:

```suggestion
      constrs0.foreach {
```

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/EffectProvenance.scala:
```diff
@@ -0,0 +1,198 @@
+              if (expected.typeVars.isEmpty && actual.typeVars.contains(effVar)) {
+                expected match {
+                  // If all actual effects are not expected, we have a path.
+                  case _ => {
```
Redundant match?

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/EffectProvenance.scala:
```diff
@@ -0,0 +1,198 @@
+  }
+
+  // Represents a path through the constraints from a source to a sink.
+  object EffectProvenancePath {
```
Will there be more than one kind of path in the future? If not, we can just turn this into a case class directly.

gaganchandan commented:
> @gaganchandan What's the status on this?

I'm working on the changes suggested by Matt

magnus-madsen commented:
> > @mlutze How can we connect the computed paths to some error message?
> 
> Maybe we can start with an unexpected effect error, where we show both endpoints of the path?

What I mean is; how does Gagan connect his code to the compiler?

mlutze commented:
> > > @mlutze How can we connect the computed paths to some error message?
> > 
> > 
> > Maybe we can start with an unexpected effect error, where we show both endpoints of the path?
> 
> What I mean is; how does Gagan connect his code to the compiler?

Maybe we can move this analysis closer to the type error reporting site?

There's somewhere we convert all the leftover constraints to errors directly.
We can probably check that the problem is an effect problem there and then create a specific error message.

gaganchandan commented:
I've cleaned up the code a little by introducing helpers and removing some unnecessary stuff. I've also introduced a new `Option`-esque type to represent source effects.

mlutze commented:
Ready for review?

gaganchandan commented:
> Ready for review?

Yes.
